src/agent.py

```diff
@@ -96,4 +101,112 @@ def _on_agent_relation_joined(self, event: RelationJoinedEvent) -> None:
         event.relation.data[self.model.unit].update(
             AgentRelationData(url=f"http://{host}:{jenkins.WEB_PORT}", secret=secret)
         )
-        self.charm.unit.status = ActiveStatus()
+        self.charm.unit.status = ops.ActiveStatus()
+
+    def _on_agent_relation_joined(self, event: ops.RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired from an agent joining the relationship.
+        """
+        container = self.charm.unit.get_container(self.state.jenkins_service_name)
+        if not container.can_connect():
+            event.defer()
+            return
+        if not event.unit or not all(
+            event.relation.data[event.unit].get(required_data)
```

Comment:
Should the relation data be part of the state and validated using pydantic?

Reply:
Refactored to be in state, thank you for the review!